File: frontend/lib/features/patient/models/medical_history.dart
// lib/features/patient/models/medical_history.dart
class MedicalHistory {
  final String? id;
  final String patientId;
  final double? bloodGlucoseLevel;
  final int? heartRate;
  final int? oxygenSaturation;
  final String? bloodPressure;
  final int? respiratoryRate;
  final double? bodyTemperature;
  final double? weight;
  final double? height;
  final List<String>? chronicDiseases;
  final List<String>? allergies;
  final String? smokingStatus;
  final String? alcoholConsumption;
  final List<CurrentMedication>? currentMedications;

  MedicalHistory({
    this.id,
    required this.patientId,
    this.bloodGlucoseLevel,
    this.heartRate,
    this.oxygenSaturation,
    this.bloodPressure,
    this.respiratoryRate,
    this.bodyTemperature,
    this.weight,
    this.height,
    this.chronicDiseases,
    this.allergies,
    this.smokingStatus,
    this.alcoholConsumption,
    this.currentMedications,
  });

  factory MedicalHistory.fromJson(Map<String, dynamic> json) {
    return MedicalHistory(
      id: json['_id'] ?? json['id'],
      patientId: json['patientId'] ?? '',
      bloodGlucoseLevel: json['bloodGlucoseLevel']?.toDouble(),
      heartRate: json['heartRate'],
      oxygenSaturation: json['oxygenSaturation'],
      bloodPressure: json['bloodPressure'],
      respiratoryRate: json['respiratoryRate'],
      bodyTemperature: json['bodyTemperature']?.toDouble(),
      weight: json['weight']?.toDouble(),
      height: json['height']?.toDouble(),
      chronicDiseases: (json['chronicDiseases'] as List?)?.cast<String>(),
      allergies: (json['allergies'] as List?)?.cast<String>(),
      smokingStatus: json['smokingStatus'],
      alcoholConsumption: json['alcoholConsumption'],
      currentMedications: (json['currentMedications'] as List?)
          ?.map((e) => CurrentMedication.fromJson(e))
          .toList(),
    );
  }

// Dans medical_history.dart, modifier la méthode toJson()
  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      if (id != null) 'id': id,
      'patientId': patientId,
    };

    // 👈 N'ajouter les champs que s'ils ne sont pas null
    if (bloodGlucoseLevel != null)
      json['bloodGlucoseLevel'] = bloodGlucoseLevel;
    if (heartRate != null) json['heartRate'] = heartRate;
    if (oxygenSaturation != null) json['oxygenSaturation'] = oxygenSaturation;
    if (bloodPressure != null && bloodPressure!.isNotEmpty)
      json['bloodPressure'] = bloodPressure;
    if (respiratoryRate != null) json['respiratoryRate'] = respiratoryRate;
    if (bodyTemperature != null) json['bodyTemperature'] = bodyTemperature;
    if (weight != null) json['weight'] = weight;
    if (height != null) json['height'] = height;

    // 👈 Toujours inclure ces champs (ils peuvent être des listes vides)
    json['chronicDiseases'] = chronicDiseases ?? [];
    json['allergies'] = allergies ?? [];

    if (smokingStatus != null) json['smokingStatus'] = smokingStatus;
    if (alcoholConsumption != null)
      json['alcoholConsumption'] = alcoholConsumption;

    json['currentMedications'] =
        currentMedications?.map((e) => e.toJson()).toList() ?? [];

    return json;
  }
}

class CurrentMedication {
  final String name;
  final String dosage;
  final String frequency;

  CurrentMedication({
    required this.name,
    required this.dosage,
    required this.frequency,
  });

  factory CurrentMedication.fromJson(Map<String, dynamic> json) {
    return CurrentMedication(
      name: json['name'] ?? '',
      dosage: json['dosage'] ?? '',
      frequency: json['frequency'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'dosage': dosage,
      'frequency': frequency,
    };
  }
}

File: services/patient-service/src/models/MedicalHistory.js
// models/MedicalHistory.js - VERSION CORRIGÉE POUR VOTRE MODÈLE IA
const mongoose = require('mongoose');

const MedicalHistorySchema = new mongoose.Schema({
  patientId: { 
    type: mongoose.Schema.Types.ObjectId, 
    ref: 'Patient', 
    required: true 
  },
  bloodGlucoseLevel: { type: Number },
  heartRate: { type: Number },
  oxygenSaturation: { type: Number },
  bloodPressure: { 
    type: mongoose.Schema.Types.Mixed, // ✅ CORRECTION: Accepter Number ou String
    validate: {
      validator: function(value) {
        if (value === null || value === undefined) return true;
        // Accepter les nombres (comme 118) ou les strings (comme "120/80")
        if (typeof value === 'number') return value > 0;
        if (typeof value === 'string') return /^\d+\/\d+$/.test(value);
        return false;
      },
      message: 'Blood pressure must be a number or format like "120/80"'
    }
  },
  respiratoryRate: { type: Number },
  bodyTemperature: { type: Number }, // In Celsius
  weight: { type: Number }, // In kg
  height: { type: Number }, // In cm
  
  // ✅ CORRECTION: Accepter String ou Array pour compatibilité modèle IA
  chronicDiseases: { 
    type: mongoose.Schema.Types.Mixed,
    validate: {
      validator: function(value) {
        if (value === null || value === undefined) return true;
        if (Array.isArray(value)) return true;
        if (typeof value === 'string') return true;
        return false;
      }
    },
    set: function(value) {
      // Convertir automatiquement string en array si nécessaire
      if (typeof value === 'string') {
        if (value === 'None' || value === '') return [];
        return value.includes(',') ? value.split(',').map(s => s.trim()) : [value];
      }
      return value;
    }
  },
  
  allergies: { 
    type: mongoose.Schema.Types.Mixed,
    validate: {
      validator: function(value) {
        if (value === null || value === undefined) return true;
        if (Array.isArray(value)) return true;
        if (typeof value === 'string') return true;
        return false;
      }
    },
    set: function(value) {
      // Convertir automatiquement string en array si nécessaire
      if (typeof value === 'string') {
        if (value === 'None' || value === '') return [];
        return value.includes(',') ? value.split(',').map(s => s.trim()) : [value];
      }
      return value;
    }
  },
  
  smokingStatus: { 
    type: String, 
    enum: ['Non-smoker', 'Ex-smoker', 'Smoker', 'Former smoker'] // ✅ AJOUT: Former smoker
  },
  
  alcoholConsumption: { 
    type: String, 
    enum: ['Never', 'Occasionally', 'Regularly', 'No'] // ✅ AJOUT: No
  },
  
  currentMedications: [{ 
    name: String,
    dosage: String,
    frequency: String 
  }]
}, { timestamps: true });

// ✅ NOUVEAU: Middleware pour nettoyer les données avant sauvegarde
MedicalHistorySchema.pre('save', function(next) {
  // Convertir bloodPressure number en string si nécessaire pour l'affichage
  if (typeof this.bloodPressure === 'number') {
    // Garder le nombre tel quel - votre modèle IA utilise des nombres
    // Pas de conversion nécessaire
  }
  
  // S'assurer que les arrays vides sont corrects
  if (!this.chronicDiseases) this.chronicDiseases = [];
  if (!this.allergies) this.allergies = [];
  if (!this.currentMedications) this.currentMedications = [];
  
  next();
});

module.exports = mongoose.model('MedicalHistory', MedicalHistorySchema);